File: Valid Expression - GFG/valid-expression.java
//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*; 
class GFG{
    public static void main(String args[]) throws IOException { 
        Scanner sc = new Scanner(System.in);
        int t = Integer.parseInt(sc.nextLine());
        while(t-- > 0){
            String S = sc.nextLine().trim();
            Solution ob = new Solution();
            if(ob.valid(S))
                System.out.println(1);
            else
                System.out.println(0);
        }
    }
}
// } Driver Code Ends


//User function Template for Java
class Solution 
{ 
    boolean valid(String s) 
    { 
        // code here
        Stack<Character> st = new Stack<>();
        for(int i=0 ; i<s.length() ; i++){
            char ch = s.charAt(i);
            if(ch=='{' || ch=='(' || ch=='['){
                st.push(ch);
            }
            else{
                if(st.isEmpty())
                    return false;
                if((ch=='}' && st.peek()=='{') || (ch==')' && st.peek()=='(') || (ch==']' && st.peek()=='[')){
                    st.pop();
                }
                else{
                    return false;
                }
            }
        }
        return st.isEmpty();
    }
} 

File: Valid Expression - GFG/valid-expression.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;
bool valid(string str);
int main()
{
    int t;
    cin>>t;
    cin.ignore();
    while(t--)
    {
        string str;
        getline(cin,str);
        cout<<valid(str)<<endl;
    }
    return 0;
}
// } Driver Code Ends


bool valid(string s)
{
    // code here
    stack<char> st;
    for(char ch: s){
        if(ch=='{' or ch=='(' or ch=='['){
            st.push(ch);
        }
        else{
            if(st.empty())
                return false;
            if((ch=='}' and st.top()=='{') or (ch==')' and st.top()=='(') or (ch==']' and st.top()=='[')){
                st.pop();
            }
            else{
                return false;
            }
        }
    }
    return st.empty();
}